### PR TITLE
rebar3: 3.4.3 -> 3.6.1

### LINKED ISSUE #38966: Update Rebar3 package
I would be nice to bump/update `Rebar3`. We currently have version `3.4.3` (checked on `Apr 15, 2018`):

```shell
[gorre@uplink ~]$ nix-env -qa ".*rebar.*"
rebar-2.5.1
rebar3-3.4.3
rebar3-3.4.3
```

...and the latest release version (`3.5.0`) was in `Dec 22, 2017`.

## Technical details

```shell
[gorre@uplink ~]$ nix-shell -p nix-info --run "nix-info -m"
these paths will be fetched (2.01 MiB download, 10.61 MiB unpacked):
  /nix/store/277x7fwjlr0mfdbdhb0ri3d8yhdk94nq-bash-4.4-p19
  /nix/store/4x22cs2ldrss83q3cdw4p89i7kb08k6n-bash-4.4-p19-dev
  /nix/store/6cnl6afcbdrg8hn80192yg211il3ixy0-bash-4.4-p19-doc
  /nix/store/6kz2vbh98s2r1pfshidkzhiy2s2qdw0a-stdenv-linux
  /nix/store/a2j35nc1m40kz0dwl76qf120lagvkcgr-readline-7.0p3
  /nix/store/afz06vlnhc6il6f4jd82c8hs9g7jq8gn-bash-4.4-p19-man
  /nix/store/d6p7343j9ywfynr23998nkqndq0fc98a-bash-4.4-p19-info
  /nix/store/gxnh7h184sbgm571242yscgkw4wy8wvx-bash-4.4-p19-dev
  /nix/store/ks166vgkssxchp5dn9hzd5njaggw4nfl-bash-4.4-p19-doc
  /nix/store/nhfgv0xxi9shl5lvl9jaf11vmm9zkv42-nix-info
  /nix/store/pa465sfqgqgnicj1ishd7n2sqapqjsq7-bash-4.4-p19-man
  /nix/store/pf6afvy0mymgmpkjc5gnk3l5ppz99mv3-bash-4.4-p19-info
copying path '/nix/store/6cnl6afcbdrg8hn80192yg211il3ixy0-bash-4.4-p19-doc' from 'https://cache.nixos.org'...
copying path '/nix/store/4x22cs2ldrss83q3cdw4p89i7kb08k6n-bash-4.4-p19-dev' from 'https://cache.nixos.org'...
copying path '/nix/store/ks166vgkssxchp5dn9hzd5njaggw4nfl-bash-4.4-p19-doc' from 'https://cache.nixos.org'...
copying path '/nix/store/d6p7343j9ywfynr23998nkqndq0fc98a-bash-4.4-p19-info' from 'https://cache.nixos.org'...
copying path '/nix/store/pf6afvy0mymgmpkjc5gnk3l5ppz99mv3-bash-4.4-p19-info' from 'https://cache.nixos.org'...
copying path '/nix/store/afz06vlnhc6il6f4jd82c8hs9g7jq8gn-bash-4.4-p19-man' from 'https://cache.nixos.org'...
copying path '/nix/store/pa465sfqgqgnicj1ishd7n2sqapqjsq7-bash-4.4-p19-man' from 'https://cache.nixos.org'...
copying path '/nix/store/nhfgv0xxi9shl5lvl9jaf11vmm9zkv42-nix-info' from 'https://cache.nixos.org'...
copying path '/nix/store/a2j35nc1m40kz0dwl76qf120lagvkcgr-readline-7.0p3' from 'https://cache.nixos.org'...
copying path '/nix/store/6kz2vbh98s2r1pfshidkzhiy2s2qdw0a-stdenv-linux' from 'https://cache.nixos.org'...
copying path '/nix/store/277x7fwjlr0mfdbdhb0ri3d8yhdk94nq-bash-4.4-p19' from 'https://cache.nixos.org'...
copying path '/nix/store/gxnh7h184sbgm571242yscgkw4wy8wvx-bash-4.4-p19-dev' from 'https://cache.nixos.org'...
 - system: `"x86_64-linux"`
 - host os: `Linux 4.15.15-1-ARCH, Arch Linux, noversion`
 - multi-user?: `no`
 - sandbox: `no`
 - version: `nix-env (Nix) 2.0`
 - channels(gorre): `"nixpkgs-18.09pre134800.e42ae4e586c"`
 - nixpkgs: `/home/gorre/.nix-defexpr/channels/nixpkgs`
```

**PS:** I didn't know how do we usually "ask" for package updates, so I file an issue O:)

### PR DESCRIPTION
**Pull request by tazjin**
Left to do:

* [x] get erlang-nix/hex-pm-registry-snapshots/#4 merged or host the registry snapshot elsewhere
* [ ] ~update `hex-packages.nix` (currently running on my machine)~ (will be a followup PR)
* [x] ensure that existing rebar-based packages still build correctly
----------------------

Updates rebar3 to version 3.6.1, which amongst other things introduces
support for rebar3 on Erlang/OTP 21.

Changes made:

* rebar3 and dependencies updated to new versions
* rebar3 hermeticity patch updated to apply against new version
* hex package registry snapshot updated

###### Motivation for this change

Nixpkgs already includes Erlang/OTP 21. The current included version of rebar3 is incompatible with the Erlang release, so an update is required.

This fixes #38966.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by gleber** [comment]
@grahamcofborg build rebar3 rebar3-open

**Comment by tazjin** [comment]
@gleber `nox-review` is showing some packages as failing (e.g. `hex2nix` itself). That package specifically has [no version bounds](https://github.com/erlang-nix/hex2nix/blob/master/rebar.config) and it seems that rebar3 will attempt to update its dependencies (`erlware-commons` in this case) to the latest version (which isn't yet in `hex-packages.nix`).

Do you know if it's expected behaviour that rebar3 wants non-constrained packages to be the latest version?

I'm currently generating an updated `hex-packages.nix`, but the process is failing every few minutes with a TLS error. Seems like it picks up again when using the `-c` flag though.

**Comment by tazjin** [comment]
Okay, for some reason this TLS connection does not succeed in Erlang 19, but it does in Erlang 20.

Does anyone know if there's a reason for the default Erlang in Nix still being at 19?

**Comment by tazjin** [comment]
Package regeneration has failed multiple times on various random issues and it seems to be difficult to resume where `hex2nix` stopped.

I'm going to see if I can fix the issues relevant for this PR manually, and then work on a package set update in a new PR. Does that sound reasonable, @gleber?

**Comment by gleber** [comment]
@tazjin This sounds good.

**Comment by tazjin** [comment]
The failing build was hex2nix itself and I've now worked around that by pinning its dependencies in the `patchPhase`. `nox-review` is now happy with this PR!

File: pkgs/development/beam-modules/hex-registry-snapshot.nix [hunk 1]
```diff
@@ -2,15 +2,14 @@
...
     name = "hex-registry";
-    rev = "9f736e7";
-    version = "0.0.0+build.${rev}";
+    rev = "11d7a24e9f53f52490ce255a6248e71128e73aa1";
+    version = "20180712.${rev}";
 
-    # src = /home/gleber/code/erl/hex-pm-registry-snapshots;
     src = fetchFromGitHub {
-        owner = "erlang-nix";
+        owner = "tazjin";
```
Comment (gleber) [inline pkgs/development/beam-modules/hex-registry-snapshot.nix]: Please revert back to `erlang-nix`

File: pkgs/development/tools/erlang/hex2nix/default.nix [hunk 2]
```diff
@@ -30,6 +30,20 @@ buildRebar3 rec {
 
     enableDebugInfo = true;
 
+    # hex2nix does not specify version constraints, in which case
```
Comment (gleber) [inline pkgs/development/tools/erlang/hex2nix/default.nix]: Maybe it's better to submit these changes directly to `hex2nix` repo? It seems like the right thing to do.

**Review by gleber — CHANGES_REQUESTED** [review]

File: pkgs/development/beam-modules/hex-registry-snapshot.nix (hunk 1 above)
Comment (tazjin) [reply pkgs/development/beam-modules/hex-registry-snapshot.nix]: Fixing in next push, good catch! :)

File: pkgs/development/tools/erlang/hex2nix/default.nix (hunk 2 above)
Comment (tazjin) [reply pkgs/development/tools/erlang/hex2nix/default.nix]: That does actually make sense, I'll prepare a change. Currently JSX is imported manually at version 2.8.2 in the derivation for hex2nix, do you know if there's a reason for it? The latest packaged version is 2.8.0 and I'd prefer to stick to that until the packages are regenerated.

Comment (tazjin) [reply pkgs/development/tools/erlang/hex2nix/default.nix]: Cursory test of hex2nix compiled with JSX 2.8.0 shows that it still works. The changelog for JSX contains nothing that looks relevant between 2.8.0 and 2.8.2.

Comment (tazjin) [reply pkgs/development/tools/erlang/hex2nix/default.nix]: Change submitted here: https://github.com/erlang-nix/hex2nix/pull/22

**Comment by tazjin** [comment]
Alright, updated the hex2nix-derivation after the merge of erlang-nix/hex2nix#22 now. I've appended the commit short hash to the version, but if it's released as a version `0.0.7` or something like that I'll bump the version to that instead.

**Review by gleber — APPROVED** [review]
Thanks for pushing it forward!

**Comment by xeji** [comment]
is this ready to merge now?

**Comment by tazjin** [comment]
@xeji Yep. I ran `nox-review` again and everything is working fine.

**Comment by xeji** [comment]
Great. Let's just check the bumped packages then since it's standard procedure:
@GrahamcOfBorg build rebar3 hex2nix

**Pull request merged**